sycl/doc/extensions/ExtendedAtomics/SYCL_INTEL_extended_atomics.asciidoc

```diff
@@ -113,389 +115,399 @@ In SYCL 1.2.1, support for floating-point types is limited to the +load+, +store
-  access::address_space::global_space>
-class atomic {
- public:
-  template <typename pointerT>
-  atomic(multi_ptr<pointerT, addressSpace> ptr);
-
-  void store(T operand, memory_order memoryOrder =
-  memory_order::relaxed);
-
-  T load(memory_order memoryOrder = memory_order::relaxed) const;
-
-  T exchange(T operand, memory_order memoryOrder =
-    memory_order::relaxed);
-
-  /* Available only when: T != float */
-  bool compare_exchange_strong(T &expected, T desired,
-    memory_order successMemoryOrder = memory_order::relaxed,
-    memory_order failMemoryOrder = memory_order::relaxed);
-
-  /* Available only when: T != float */
-  T fetch_add(T operand, memory_order memoryOrder =
-    memory_order::relaxed);
-
-  /* Available only when: T != float */
-  T fetch_sub(T operand, memory_order memoryOrder =
-    memory_order::relaxed);
-
-  /* Available only when: T != float */
-  T fetch_and(T operand, memory_order memoryOrder =
-    memory_order::relaxed);
-
-  /* Available only when: T != float */
-  T fetch_or(T operand, memory_order memoryOrder =
-    memory_order::relaxed);
-
-  /* Available only when: T != float */
-  T fetch_xor(T operand, memory_order memoryOrder =
-    memory_order::relaxed);
-
-  /* Available only when: T != float */
-  T fetch_min(T operand, memory_order memoryOrder =
-    memory_order::relaxed);
-
-  /* Available only when: T != float */
-  T fetch_max(T operand, memory_order memoryOrder =
-    memory_order::relaxed);
+inline constexpr memory_order memory_order_relaxed = memory_order::relaxed;
+inline constexpr memory_order memory_order_acquire = memory_order::acquire;
+inline constexpr memory_order memory_order_release = memory_order::release;
+inline constexpr memory_order memory_order_acq_rel = memory_order::acq_rel;
+inline constexpr memory_order memory_order_seq_cst = memory_order::seq_cst;
+
+// Exposition only
+template <memory_order ReadModifyWriteOrder>
+struct memory_order_traits;
+
+template <>
+struct memory_order_traits<memory_order::relaxed> {
+  static constexpr memory_order read_order = memory_order::relaxed;
+  static constexpr memory_order write_order = memory_order::relaxed;
 };
-} // namespace sycl
-} // namespace cl
-----
 
-==== To:
+template <>
+struct memory_order_traits<memory_order::acq_rel> {
+  static constexpr memory_order read_order = memory_order::acquire;
+  static constexpr memory_order write_order = memory_order::release;
+};
 
-[source,c++]
-----
-namespace cl {
-namespace sycl {
-enum class memory_order : int {
-  relaxed
+template <>
+struct memory_order_traits<memory_order::seq_cst> {
+  static constexpr memory_order read_order = memory_order::seq_cst;
+  static constexpr memory_order write_order = memory_order::seq_cst;
 };
-namespace intel {
-template <typename T, access::address_space addressSpace =
-  access::address_space::global_space>
-class atomic {
+
+template <typename T, memory_order DefaultOrder, access::address_space Space>
```

Comment:
What about default value for `Space`? Also, I guess there are some limitations, i.e. it doesn't make sense to create `atomic_ref` to constant or private memory

Reply:
Like with `multi_ptr` and the existing `atomic`, I don't see a good way to default the value of `Space`.  Nothing in the compiler that I can see currently checks whether a `T*` is a global or local pointer, so I think we'd just end up with lots of run-time failures.

Reply:
I agree about the limitations, though, and I've added a fix for that.